### DISCUSSION TITLE
Plugin config fields contain the same values

### DISCUSSION THREAD

**Discussion by vanlueckn** [kimai/kimai#3289] [General]
I am programming a plugin to export invoices to SevDesk.

I have also added my own settings for this. When I enter different values in the settings and save them, all fields suddenly have the same content:

![brave_tWdOPQCoUW](https://user-images.githubusercontent.com/53119369/166685945-40f48477-638e-477a-b3d2-969e91d08ef1.png)

My Config Class:

```php
<?php

namespace KimaiPlugin\KimaiSevdeskBundle\Configuration;

use App\Configuration\SystemConfiguration;

final class SevdeskConfiguration
{
    private SystemConfiguration $configuration;

    public function __construct(SystemConfiguration $configuration)
    {
        $this->configuration = $configuration;
    }

    public function getSevdeskApiKey(): string
    {
        return (string)$this->configuration->find('sevdesk.api_key');
    }

    public function getSevdeskContactPersonId(): int
    {
        return (int)$this->configuration->find('sevdesk.contact_person_id');
    }

    public function getTaxRate(): int
    {
        return (int)$this->configuration->find('sevdesk.tax_rate');
    }

    public function getSevdeskContactId(): int
    {
        return (int)$this->configuration->find('sevdesk.contact_id');
    }

    public function getSevdeskHourUnitId(): int
    {
        return (int)$this->configuration->find('sevdesk.hour_unit_id');
    }

}
```

My EventSubscriber:

```php
<?php

namespace KimaiPlugin\KimaiSevdeskBundle\EventSubscriber;

use App\Event\SystemConfigurationEvent;
use App\Form\Model\Configuration;
use App\Form\Model\SystemConfiguration as SystemConfigurationModel;
use Symfony\Component\EventDispatcher\EventSubscriberInterface;
use Symfony\Component\Form\Extension\Core\Type\TextType;

class SystemConfigurationSubscriber implements EventSubscriberInterface
{
    public static function getSubscribedEvents()
    {
        return [
            SystemConfigurationEvent::class => ['onSystemConfiguration', 100],
        ];
    }

    public function onSystemConfiguration(SystemConfigurationEvent $event)
    {
        $event->addConfiguration(
            (new SystemConfigurationModel())
                ->setSection('sevdesk_config')
                ->setConfiguration([
                    (new Configuration())
                        ->setName('sevdesk.api_key')
                        ->setLabel('sevdesk.api_key')
                        ->setTranslationDomain('system-configuration')
                        ->setRequired(false)
                        ->setType(TextType::class),
                    (new Configuration())
                        ->setName('sevdesk.contact_person_id')
                        ->setLabel('sevdesk.contact_person_id')
                        ->setTranslationDomain('system-configuration')
                        ->setRequired(false)
                        ->setType(TextType::class),
                    (new Configuration())
                        ->setName('sevdesk.tax_rate')
                        ->setLabel('sevdesk.tax_rate')
                        ->setTranslationDomain('system-configuration')
                        ->setRequired(false)
                        ->setType(TextType::class),
                    (new Configuration())
                        ->setName('sevdesk.contact_id')
                        ->setLabel('sevdesk.contact_id')
                        ->setTranslationDomain('system-configuration')
                        ->setRequired(false)
                        ->setType(TextType::class),
                    (new Configuration())
                        ->setName('sevdesk.hour_unit_id')
                        ->setLabel('sevdesk.hour_unit_id')
                        ->setTranslationDomain('system-configuration')
                        ->setRequired(false)
                        ->setType(TextType::class),
                ])
        );
    }
}
```

My Configuration Class:

```php
<?php

namespace KimaiPlugin\KimaiSevdeskBundle\DependencyInjection;

use Symfony\Component\Config\Definition\Builder\ArrayNodeDefinition;
use Symfony\Component\Config\Definition\Builder\TreeBuilder;
use Symfony\Component\Config\Definition\ConfigurationInterface;

class Configuration implements ConfigurationInterface
{
    /**
     * {@inheritdoc}
     */
    public function getConfigTreeBuilder(): TreeBuilder
    {
        $treeBuilder = new TreeBuilder('sevdesk');
        /** @var ArrayNodeDefinition $rootNode */
        $rootNode = $treeBuilder->getRootNode();

        $rootNode
            ->addDefaultsIfNotSet()
            ->children()
            ->scalarNode('api_key')
            ->defaultValue('REPLACE')
            ->end()
            ->scalarNode('contact_person_id')
            ->defaultValue('0')
            ->end()
            ->scalarNode('tax_rate')
            ->defaultValue('19')
            ->end()
            ->scalarNode('contact_id')
            ->defaultValue('0')
            ->end()
            ->scalarNode('hour_unit_id')
            ->defaultValue('9')
            ->end()
            ->end()
            ->end();

        return $treeBuilder;
    }
}
```

My AbstractPluginExtension:

```php
<?php

namespace KimaiPlugin\KimaiSevdeskBundle\DependencyInjection;

use App\Plugin\AbstractPluginExtension;
use Symfony\Component\Config\FileLocator;
use Symfony\Component\DependencyInjection\ContainerBuilder;
use Symfony\Component\DependencyInjection\Loader;

class KimaiSevdeskExtension extends AbstractPluginExtension
{
    public function load(array $configs, ContainerBuilder $container)
    {
        $configuration = new Configuration();
        $config = $this->processConfiguration($configuration, $configs);
        $this->registerBundleConfiguration($container, $config);

        $loader = new Loader\YamlFileLoader($container, new FileLocator(__DIR__ . '/../Resources/config'));
        $loader->load('services.yaml');
    }
}
```

**Comment by kevinpapst** [comment]
Sounds wrong. No idea, that is basically what I use in all plugins.
BTW. You don't need the `Configuration` class or the code in the `KimaiSevdeskExtension`... only if you want to support configuration via local.yaml. The SystemConfiguration class works totally fine without that definition.
Are the values also in the database wrong?

**Reply by vanlueckn** [reply]
I removed the unnecessary classes.

The values in the databased are correct:

![MobaXterm_2Uqt1GYb2s](https://user-images.githubusercontent.com/53119369/166922748-6c26ba58-0fa4-402e-84a5-75c75b21eea3.png)

The values in the ui before saving;

![brave_S7n7inyWIR](https://user-images.githubusercontent.com/53119369/166923025-49cdcdc7-bfd1-4516-ab2e-c42bd390c54e.png)

The values in the ui after saving:

![brave_tWdOPQCoUW](https://user-images.githubusercontent.com/53119369/166922871-39c47a45-e305-48b8-8253-204b3dac87f5.png)

**Reply by kevinpapst** [reply]
And then after page reload? Again the correct values?
Did you try to dump the configuration and the result of find()? What was the result?

**Reply by vanlueckn** [reply]
Even after a reload, the values remain the same. Also my SevdeskConfiguration.php class / find() always returns the same values.

**Reply by vanlueckn** [reply]
So the database contains the correct values but find() returns 5 for every field.

**Reply by vanlueckn** [reply]
I pushed the code to github: https://github.com/vanlueckn/kimai-sevdesk-plugin

**Reply by kevinpapst** [reply]
I just dumped your `onSystemConfiguration()` code into one of my plugins and this happens:

First result is entirely empty (expected):
<img width="555" alt="Bildschirmfoto 2022-05-05 um 14 47 28" src="https://user-images.githubusercontent.com/533162/166926259-7d3e535e-d130-4003-bdd9-64fefc7283b4.png">

After saving and then also after page reload all good:
<img width="299" alt="Bildschirmfoto 2022-05-05 um 14 47 45" src="https://user-images.githubusercontent.com/533162/166926325-eab9a6a6-f9a2-4777-8abd-bb77f5283120.png">

Database obviously the same:
<img width="455" alt="Bildschirmfoto 2022-05-05 um 14 48 02" src="https://user-images.githubusercontent.com/533162/166926374-4cd85d26-d8a5-45bf-942a-1f6025b9454e.png">

Must be something else in your code.

**Reply by kevinpapst** [reply]
I installed your code, works the same and as expected: correct.
<img width="360" alt="Bildschirmfoto 2022-05-05 um 14 53 36" src="https://user-images.githubusercontent.com/533162/166927101-99ed5dd5-f6a1-4023-9e0e-0377c797e42d.png">

And after save & reload
<img width="278" alt="Bildschirmfoto 2022-05-05 um 14 53 54" src="https://user-images.githubusercontent.com/533162/166927115-72b0673d-c1a7-4abe-8777-2c1b5b3dbf65.png">

Sorry, must be something else. Your code is working fine.

Which Kimai version? Which PHP version? 
What does `git status` report?
Do you have composer changes?

**Reply by vanlueckn** [reply]
Version: 1.18.1 stable / Kimai Docker Image
PHP 8.0.15
git status is clean
no composer changes

**Reply by vanlueckn** [reply]
my installed plugins
![brave_Ojhsl3pXJy](https://user-images.githubusercontent.com/53119369/166933510-07a827c1-f89f-4013-ab3b-380c69d5a28b.png)

**Reply by kevinpapst** [reply]
Please upgrade to latest release before further investigations

**Reply by kevinpapst** [reply]
Also: please run in dev mode with activated debugging and dev composer packages. Docker is not ideal for that development...

**Reply by vanlueckn** [reply]
Upgrading to the latest version solved the issue, thank you
